src/unix/process.c

```diff
@@ -342,6 +345,181 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  }
+
+  /* Set flags for spawn behavior 
+    * 1) POSIX_SPAWN_CLOEXEC_DEFAULT: (Apple Extension) All descriptors in
+    * the parent will be treated as if they had been created with O_CLOEXEC. 
+    * The only fds that will be passed on to the child are those manipulated by the file actions
+    * 2) POSIX_SPAWN_SETSIGDEF: Signals mentioned in spawn-sigdefault in the spawn attributes
+    * will be reset to behave as their default
+    * 3) POSIX_SPAWN_SETSIGMASK: Signal mask will be set to the value of spawn-sigmask in attributes
+    * 4) POSIX_SPAWN_SETSID: Make the process a new session leader if a detached session was requested. */
+  flags = POSIX_SPAWN_CLOEXEC_DEFAULT |
+    POSIX_SPAWN_SETSIGDEF |
+    POSIX_SPAWN_SETSIGMASK;
+  if(options->flags & UV_PROCESS_DETACHED) 
+    flags |= POSIX_SPAWN_SETSID;
+  err = posix_spawnattr_setflags(&attrs, flags);
+  if(err != 0)
+    return err;
+
+  // Reset all signal the child to their default behavior
+  sigfillset(&signal_set);
+  err = posix_spawnattr_setsigdefault(&attrs, &signal_set);
+  if(err != 0) 
+    return err;
+
+  // Reset the signal mask for all signals
+  sigemptyset(&signal_set);
+  err = posix_spawnattr_setsigmask(&attrs, &signal_set);
+  if(err != 0)
+    return err;
+
+  err = posix_spawn_file_actions_init(&actions);
+  if(err != 0)
+    return err;
+
+  if (options->cwd != NULL && posix_spawn_file_actions_addchdir_np(&actions, options->cwd))
+    return err;
+
+  // First, dupe any required fd into orbit, out of the range of 
+  // the descriptors that should be mapped in.
+  for(fd = 0 ; fd < stdio_count; ++fd) {
+    if(pipes[fd][1] < 0)
+      continue;
+    
+    err = posix_spawn_file_actions_adddup2(&actions, pipes[fd][1], stdio_count + fd);
+    if(err != 0)
+      return err;
+  }
+
+  // Second, move the descriptors into their respective places
+  for(fd = 0 ; fd < stdio_count; ++fd) {
+    if(pipes[fd][1] < 0)
+      continue;
+
+    err = posix_spawn_file_actions_adddup2(&actions, stdio_count + fd, fd);
+    if(err != 0)
+      return err;
+  }
+
+  // Finally, close all the superfluous descriptors
+  for(fd = 0; fd < stdio_count; ++fd) {
+    if(pipes[fd][1] < 0)
+      continue;
+    
+    err = posix_spawn_file_actions_addclose(&actions, stdio_count + fd);
+    if(err != 0)
+      return err;
+  }
+
+  // Finally process the standard streams as per de documentation
+  for(fd = 0 ; fd < 3 ; ++fd) {
+    // If ignored, open as /dev/null
+    const int oflags = fd == 0 ? O_RDONLY : O_RDWR;
+    const int mode = 0;
+
+    if(pipes[fd][1] != -1) 
+      continue;
+    
+    err = posix_spawn_file_actions_addopen(&actions, fd, "/dev/null", oflags, mode);
+    if(err != 0)
+      return err;
+  } 
+
+  // Preserve parent environment if not explicitly set
+  char ** env = options->env ? options->env : environ;
+
+  // Spawn the child
+  err = posix_spawnp(pid, options->file, &actions, &attrs, options->args, env);
+
+  return UV__ERR(err);
```

Comment:
Resource leak: neither the happy path nor the error paths call `posix_spawnattr_destroy()`.